// textlint/lib/linter/linter-processor.js
// LICENSE : MIT
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _assert = require("assert");

var _assert2 = _interopRequireDefault(_assert);

var _linterTask = require("../task/linter-task");

var _linterTask2 = _interopRequireDefault(_linterTask);

var _taskRunner = require("../task/task-runner");

var _taskRunner2 = _interopRequireDefault(_taskRunner);

var _messageFilter = require("../shared/message-filter");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var LinterProcessor = function () {
    /**
     *
     * @param {Processor} processor
     */

    function LinterProcessor(processor) {
        _classCallCheck(this, LinterProcessor);

        this.processor = processor;
    }

    /**
     * Run linter process
     * @param {Config} config
     * @param {RuleCreatorSet} ruleCreatorSet
     * @param {SourceCode} sourceCode
     * @returns {Promise.<TextLintResult>}
     */


    _createClass(LinterProcessor, [{
        key: "process",
        value: function process(_ref) {
            var config = _ref.config;
            var ruleCreatorSet = _ref.ruleCreatorSet;
            var sourceCode = _ref.sourceCode;

            (0, _assert2.default)(config && ruleCreatorSet && sourceCode);

            var _processor$processor = this.processor.processor(sourceCode.ext);

            var preProcess = _processor$processor.preProcess;
            var postProcess = _processor$processor.postProcess;

            (0, _assert2.default)(typeof preProcess === "function" && typeof postProcess === "function", "processor should implement {preProcess, postProcess}");
            var task = new _linterTask2.default({
                config: config,
                ruleCreatorSet: ruleCreatorSet,
                sourceCode: sourceCode
            });
            return _taskRunner2.default.process(task).then(function (messages) {
                var result = postProcess(messages, sourceCode.filePath);
                result.messages = (0, _messageFilter.filterMessages)(result.messages);
                if (result.filePath == null) {
                    result.filePath = "<Unkown" + sourceCode.ext + ">";
                }
                (0, _assert2.default)(result.filePath && result.messages.length >= 0, "postProcess should return { messages, filePath } ");
                return result;
            });
        }
    }]);

    return LinterProcessor;
}();

exports.default = LinterProcessor;
module.exports = exports['default'];
