// textlint-plugin-text/lib/TextProcessor.js
// LICENSE : MIT
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var _txtToAst = require("txt-to-ast");

var TextProcessor = (function () {
    function TextProcessor(config) {
        _classCallCheck(this, TextProcessor);

        this.config = config;
    }

    _createClass(TextProcessor, [{
        key: "processor",
        value: function processor(ext) {
            return {
                preProcess: function preProcess(text, filePath) {
                    return (0, _txtToAst.parse)(text);
                },
                postProcess: function postProcess(messages, filePath) {
                    return {
                        messages: messages,
                        filePath: filePath ? filePath : "<text>"
                    };
                }
            };
        }
    }], [{
        key: "availableExtensions",
        value: function availableExtensions() {
            return [".txt", ".text"];
        }
    }]);

    return TextProcessor;
})();

exports["default"] = TextProcessor;
module.exports = exports["default"];
